File: dsa/li05_binarysearch/java/L2387_MedianOfARowWiseSortedMatrix.java

public class L2387_MedianOfARowWiseSortedMatrix {
    public static void main(String[] args) {
        L2387_MedianOfARowWiseSortedMatrix obj = new L2387_MedianOfARowWiseSortedMatrix();

        int[][] arr = { { 1, 3, 8 },    // 1,1,2,2,3,3,4,5,8
                        { 2, 3, 4 },
                        { 1, 2, 5 } };
        System.out.println(obj.findMedian(arr));

    }

    // Problem: https://leetcode.ca/2022-10-06-2387-Median-of-a-Row-Wise-Sorted-Matrix/
    // Idea: BS range is min and max of matrix elements
    // Find the mid of min and max and count number of elements smaller than or equal mid
    // Time: O(32 * r * log(c))
    public int findMedian(int[][] mat) {
        int row = mat.length, col = mat[0].length;
        int low = Integer.MAX_VALUE, high = Integer.MIN_VALUE;
        for (int i = 0; i < row; i++) {
            low = Math.min(low, mat[i][0]);
            high = Math.max(high, mat[i][col - 1]);
        }

        while (low <= high) {
            int mid = (low + high) / 2;
            int cnt = smallerThenOrEqualMid(mat, row, mid);
            if (cnt <= (row * col)/2)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return low;
    }

    private int smallerThenOrEqualMid(int[][] mat, int row, int mid) {
        int cnt = 0;
        for (int i = 0; i < row; i++) {
            cnt += upperBound(mat[i], mid);
        }
        return cnt;
    }

    private int upperBound(int[] nums, int target) {
        int low = 0, high = nums.length - 1;
        while (low <= high) {
            int md = (low + high) / 2;
            if (nums[md] <= target) {
                low = md + 1;
            } else {
                high = md - 1;
            }
        }
        return low;
    }
}

File: dsa/concepts/05_binarysearch/python/l2387_median_of_arow_wise_sorted_matrix.py
"""
Median of a Row-Wise Sorted Matrix

Idea:
Binary search on the value range [min_element, max_element].
For a guessed value mid, count how many elements are <= mid
using upper bound on each row.

Time Complexity: O(log(range) * r * log(c))
Space Complexity: O(1)
"""

def upper_bound(nums, target):
    low, high = 0, len(nums) - 1
    while low <= high:
        mid = (low + high) // 2
        if nums[mid] <= target:
            low = mid + 1
        else:
            high = mid - 1
    return low


def count_smaller_equal(mat, mid):
    count = 0
    for row in mat:
        count += upper_bound(row, mid)
    return count


def find_median(mat):
    r, c = len(mat), len(mat[0])

    # Determine search range
    low = min(row[0] for row in mat)
    high = max(row[-1] for row in mat)

    desired = (r * c) // 2

    while low <= high:
        mid = (low + high) // 2
        cnt = count_smaller_equal(mat, mid)

        if cnt <= desired:
            low = mid + 1
        else:
            high = mid - 1

    return low


# Driver code
if __name__ == "__main__":
    mat = [
        [1, 3, 8],
        [2, 3, 4],
        [1, 2, 5]
    ]
    print(find_median(mat))  # 3
